Port FSDP::summon_full_params from fairscale to pytorch.

kumpera opened the pull request:
Summary:
Bring FSDP::summon_full_params from fairscale.

It doesn't support summoning with full precision as this mode is not yet supported by other parts of PT's FSDP port.

One thing that needs figuring out is the semantics we want W.R.T. reshard_after_forward. Right now I'm always discarding the full tensor at the end of _summon_full_params.

Fixes: https://github.com/pytorch/pytorch/issues/69779

Test Plan: Ported the fairscale tests plus added a few more.

Differential Revision: D33350378



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+        Args:
+            recurse (bool, Optional): recursively summon all params for nested
+                FSDP instances (default: True)
+            volatile (bool, Optional): if ``True``, modifications to params are
```
would just changing this arg to something like `writeback` be more clear?

kumpera replied:
I agree that `writeback` is a clearer. I was not sure about changing names in the first place.

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+        A context manager to expose full params for the current FSDP instance.
+        Can be useful *after* forward/backward for a model to get the params for
+        additional processing or checking.
+        .. note:: This can be used on inner FSDPs.
```
For the sake of understanding, does this mean that if I have:
```
FSDP(
nn.Linear,
FSDP(nn.Linear),
)
```

if I use it on the nested FSDP, I'll get only the nested linear layer's parameters? Do we have the appropriate unittest for this?

kumpera replied:
Yup, that's what it means. We don't have a test specifically for that, I'm adding a test case for that.

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+        additional processing or checking.
+        .. note:: This can be used on inner FSDPs.
+        .. note:: This can *not* be used within a forward or backward pass. Nor
+            can forward and backward be started from within this context.
```
is it worth adding a test case to verify we throw an appropriate error in this case (or file a follow up issue to do so, doesn't have to be in this PR)?

kumpera replied:
good point. the test I added is a bit wacky in how it allows `_summon_full_params` to be called from forward, but it does the job.

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+                if volatile is False:
+                    self._write_back_current_shard()
+                # FIXME should we respect #self.reshard_after_forward here too?
+                self._free_full_params()
```
Is there a reason we can't just simply check the flag here?

kumpera replied:
I was not sure on the expected behavior. It looks like checking it is the appropriate behavior.

kumpera replied:
Looks like we should check for it. Fixed and added a test.

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+            try:
+                yield
+            finally:
+                if volatile is False:
```
nit: `if not volatile`?

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+        self.assertEqual(p[0], self.rank + 2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_with_writeback(self):
```
this test seems quite similar to the one above and only seems to differ in the `volatile` param and final assert statement.

You can consider using `@parametrize` decorator to write only one unittest and parametrize it with `volatile` flag. Example of parametrized tests are in `test_fsdp_core.py`.

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+
+        model = FSDP(raw_model.cuda(self.rank))
+
+        # params should be half as many
```
Is it more accurate to say params should be 121 / world size as we can have world size > 2?

kumpera replied:
yup. I changed the test to be world_size agnostic.

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+            slice_end = (self.rank + 1) * expected_shard_size
+            my_slice = all_shards[slice_start : slice_end]
+            # shards are padded but the full_param tensor is not
+            if self.rank == self.world_size - 1:
```
wondering why they are different padding on different ranks?

kumpera replied:
Say you shard a 10 elements tensor over 3 workers.
You end up with ranks 4 element tensors on each rank.
Ranks 0 and 1 have zero padding while rank 2 has 2 elements of padding.

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+        with model._summon_full_params(recurse=False):
+            outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+            inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+            self.assertEqual(18, outer_param.numel())
```
Assuming its 18 because 9 is the shard size and world size is 2, can we generalize it to world size > 2? I am not sure if these tests run with world size > 2 but they definitely should if they don't

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+        with model._summon_full_params(recurse=True):
+            outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+            inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+            self.assertEqual(18, outer_param.numel())
```
same comment as above

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+            self.assertEqual(15, inner_param.numel())
+
+        # recursive, should summon everything
+        with model._summon_full_params(recurse=True):
```
can we test recursive summoning + write_back?

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+        self.assertEqual(9, outer_param.numel())
+        self.assertEqual(15, inner_param.numel())
+
+        # non recursive, should only summon outer
```
can we test non-recursive, but we only summon an _inner_ layer's full params?

kumpera replied:
done.

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,133 @@
+                self.assertTrue(torch.equal(my_shard, my_slice))
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_recursive(self):
```
Might be useful to test all these with `CPUOffload` set to true as well to ensure CPU offloading + summon full params work well together. You can use `parametrize` decorator to easily generate those tests.

kumpera replied:
done.

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+                pass
+            chunks = p._full_param_padded.chunk(self.world_size)
+            if len(chunks) < (self.rank + 1):
+                # if there are not enough chunks, there's nothing to writeback here
```
in which cases can this happen?

kumpera replied:
If there are more workers than elements in the tensor we run into this situation.

zhaojuanmao replied:
possibly we should not support this use case? data is already not what users expected, right?

kumpera replied:
@zhaojuanmao this behavior is supported by FSDP and is currently used by the test suite. test_fsdp_overlap.py module `Layer` has a one element parameter that will trigger this behavior. I C&P that code from `_get_shard` but have not added a test covering it.

zhaojuanmao replied:
when Layer only has one element, I thought _full_param_padded has already been padded for every rank. So this case should not be triggered at all here...?

kumpera replied:
You're right, `_full_param_padded` is already padded so that condition will never happen.

rohan-varma reviewed (commented):
Awesome, looks great! API design looks good, have some questions around tests.

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,176 @@
+        self.assertEqual(shard_inner_numel, inner_param.numel())
+
+        # non recursive, should only summon outer
+        with model._summon_full_params(recurse=False):
```
nit: can reduce duplication a bit:

```
for recurse in [true, false]:
with summon_full_params(recurse):
     self.assertEqual(global_outer_numel, outer_param.numel())
     self.assertEqual(shard_inner_numel, inner_param.numel())
```

kumpera replied:
I changed the test to use `@parametrize`

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,176 @@
+
+        model = FSDP(MyModule()).cuda(self.rank)
+        with self.assertRaisesRegex(ValueError, "current state is TrainingState_.FORWARD"):
+            model.forward(model)
```
Usually to run the forward pass we want to just call `model(inputs)` as opposed to explicit call to `model.forward()`.

This is because `__call__` implementation does some other stuff (i.e. call hooks) prior to calling into forward and usually direct call to .forward() is discouraged in my experience.

kumpera replied:
Good to know! Fixed.

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,176 @@
+                    pass
+
+        model = FSDP(MyModule()).cuda(self.rank)
+        with self.assertRaisesRegex(ValueError, "current state is TrainingState_.FORWARD"):
```
also test what happens in backward?

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+        .. note:: This can be used on inner FSDPs.
+        .. note:: This can *not* be used within a forward or backward pass. Nor
+            can forward and backward be started from within this context.
+        .. note:: The full parameters will be freed after the context manager
```
This isn't true for the root module right? As we respect `reshard_after_forward`?

kumpera replied:
Good point, they are not freed but we switch to the local shard.

rohan-varma reviewed (approved):
Looks great, thanks! Just a few more minor comments for your consideration, also please fix lint. @zhaojuanmao might also want to have a look.

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+            self.training_state = TrainingState_.SUMMON_FULL_PARAMS
+
+            self._rebuild_full_params()
+
```
we need to add the following here as well? it is to unflatten the flat_param back to original params: 

                    stack.enter_context(
                        self.module.unflatten_params(
                            flat_params=[p.data for p in self.params[: self._num_flatten_params]]
                        )
                    )

kumpera replied:
Pytorch's version of FSDP uses a single flattened parameter so we don't need unflattening

zhaojuanmao replied:
the unflattening is still needed for the single flatten parameter. The unflatten is to unflatten the single flatten parameter to original parameters, so that users can access the original unflatten parameters under this context

zhaojuanmao replied:
Thanks for addressing all other comments. This is the only left one now. 

Possibly it is good to add unflatten_params() context manager in another PR, and let this PR stack on top of it?

kumpera replied:
Fixed.

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+        for p in self.params:
+            if not p._is_sharded:
+                pass
+            chunks = p._full_param_padded.chunk(self.world_size)
```
wondering why not calling _get_shard() directly?

kumpera replied:
You're right that no padding will happen if call `self._get_shard(p._full_param_padded)`.

The issue around copying still there. In https://github.com/pytorch/pytorch/blob/master/torch/distributed/_fsdp/fully_sharded_data_parallel.py#L364 you can see `_get_shard` calls `chunk.clone()` unconditionally on the chunk.

zhaojuanmao replied:
good point, the change here looks good, thanks!

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,71 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+            finally:
+                if writeback:
+                    self._write_back_current_shard()
+                if self.reshard_after_forward:
+                    self._free_full_params()
```
we can remove reshard_after_forward check here, as reshard_after_forward is used inside an iteration (after forward, before backward). outside the forward/backward iteration, only shards are kept. so it is always safe to free full params in this context

kumpera replied:
Won't that break using `_summon_full_params` between forward and backward?
Abetter behavior would be to leave it as it was at the beginning of the call.

zhaojuanmao replied:
if this call is used after backward, self.reshard_after_forward will still be False for root module, in this case, root module's full params will not be freed, which is not expected.

kumpera replied:
I added a test for what I believe is the behavior you're describing: `test_reshard_outside_forward_backward_iteration`.
Does it make sense?

zhaojuanmao reviewed (changes requested):
great work! thanks for adding comprehensive tests! having a few comments

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +631,75 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+            finally:
+                if writeback:
+                    self._write_back_current_shard()
+                self._free_full_params(currently_local_params)
```
Thanks for the change! it covered both cases, test_reshard_outside_forward_backward_iteration() looks great!

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,264 @@
+        my_shard = torch.clone(next(model.parameters()))
+
+        with model._summon_full_params():
+            self.assertEqual(raw_model_size, self.get_model_param_count(model))
```
should we have some checking such that we ensure the param size is (10, 11) as expected?

kumpera replied:
This is done by how `raw_model_size` is computed by counting the size of a non-FSDP model.

zhaojuanmao commented:
let's address the one left comment in the follow up PR, approving to unblock. Thanks for the great work!

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,278 @@
+        with model._summon_full_params():
+            a = model.weight.flatten().detach()
+            b = flattened_param.detach()
+            self.assertTrue(torch.equal(a, b))
```
Can be addressed in a follow up diff - I think this tests that if we flatten the params they are equal, but we should ideally also test the original shapes.

rohan-varma reviewed (commented):
Quick note mostly just for myself - we should agree on what `model.parameters()` should mean outside and inside of `summon_full_params` context, as this will also affect checkpoint.

Right now `.parameters()` returns only the flat param, which means state_dict will be something like 

```
{
`flat_param`: tensor[....]
}
```

which further means it cannot be easily loaded by non-FSDP instances, though that is probably an acceptable constraint. cc @zhaojuanmao

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -629,6 +632,76 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+    def _write_back_current_shard(self):
+        for p in self.params:
+            if not p._is_sharded:  # type: ignore[attr-defined]
+                pass
```
s/pass/continue? Also I am not sure if this is covered by the tests, as I think with `pass` this can throw an error if the param is not sharded since we attempt to access `_full_param_padded` which is not set.

kumpera committed:
Port FSDP::summon_full_params from fairscale to pytorch. (#71225)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71225

Bring FSDP::summon_full_params from fairscale.

It doesn't support summoning with full precision as this mode is not yet supported by other parts of PT's FSDP port.

One thing that needs figuring out is the semantics we want W.R.T. reshard_after_forward. Right now I'm always discarding the full tensor at the end of _summon_full_params.

Fixes: https://github.com/pytorch/pytorch/issues/69779

Test Plan: Ported the fairscale tests plus added a few more.

Reviewed By: zhaojuanmao, rohan-varma

Differential Revision: D33350378

fbshipit-source-id: 7012390d2019228a101e632938be208627dcddf8

zhaojuanmao commented:
> Quick note mostly just for myself - we should agree on what `model.parameters()` should mean outside and inside of `summon_full_params` context, as this will also affect checkpoint.
> 
> Right now `.parameters()` returns only the flat param, which means state_dict will be something like
> 
> ```
> {
> `flat_param`: tensor[....]
> }
> ```
> 
> which further means it cannot be easily loaded by non-FSDP instances, though that is probably an acceptable constraint. cc @zhaojuanmao

From usability perspective, preferring to make .parameters() and .named_parameters() to return keys with original parameters names, it could be sharded data if outside summon_full_param() context and full data if inside summon_full_param()

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -0,0 +1,278 @@
+        model = FSDP(nn.Sequential(
+            FSDP(nn.Linear(5, 5, bias=False)),
+            nn.Linear(5, 3, bias=False)
+        )).cuda(self.rank)
```
Missed this in code review, but it does not actually respect the cpu_offload argument.
